File: implementations/python/dice_lab.py
from __future__ import annotations

import argparse
import csv
import json
import math
import random
import sys
from dataclasses import dataclass
from io import StringIO
from typing import Sequence


@dataclass(frozen=True)
class SimulationResult:
    total_rolls: int
    sides: int
    distribution: dict[int, int]
    mean: float
    variance: float
    std_dev: float


def parse_args(argv: Sequence[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        prog="dice-lab",
        description="Simulate dice rolls and report distribution + statistics.",
    )

    parser.add_argument(
        "--rolls",
        type=int,
        required=True,
        help="Total number of rolls to simulate (must be > 0).",
    )
    parser.add_argument(
        "--sides",
        type=int,
        default=6,
        help="Number of die sides (must be >= 2). Default: 6.",
    )
    parser.add_argument(
        "--seed",
        type=int,
        default=None,
        help="Optional seed for deterministic RNG behavior.",
    )
    parser.add_argument(
        "--format",
        choices=("text", "json", "csv"),
        default="text",
        help="Output format.",
    )
    parser.add_argument(
        "--parallel",
        action="store_true",
        help="Requested parallel execution (not implemented in Python baseline).",
    )

    args = parser.parse_args(argv)

    if args.rolls <= 0:
        raise ValueError("--rolls must be greater than 0.")
    if args.sides < 2:
        raise ValueError("--sides must be greater than or equal to 2.")

    return args


def run_simulation(rolls: int, sides: int, seed: int | None) -> SimulationResult:
    # Use a local Random instance so seeding here does not affect global random state.
    rng = random.Random(seed)

    # Pre-initialize counts for every face so output is deterministic and complete.
    counts = {face: 0 for face in range(1, sides + 1)}

    # We accumulate sum and sum-of-squares on the fly to avoid storing every roll.
    total = 0.0
    total_sq = 0.0

    for _ in range(rolls):
        value = rng.randint(1, sides)
        counts[value] += 1
        total += value
        total_sq += value * value

    mean = total / rolls

    # Population variance definition from spec: Sum((x - mean)^2) / N
    # Algebraically equivalent and efficient form using E[x^2] - (E[x])^2.
    variance = (total_sq / rolls) - (mean * mean)

    # Tiny negatives can occur from floating point rounding; clamp to zero.
    variance = max(0.0, variance)
    std_dev = math.sqrt(variance)

    return SimulationResult(
        total_rolls=rolls,
        sides=sides,
        distribution=counts,
        mean=mean,
        variance=variance,
        std_dev=std_dev,
    )


def format_text(result: SimulationResult) -> str:
    lines: list[str] = []
    lines.append("DiceLab Results")
    lines.append(f"Total rolls: {result.total_rolls}")
    lines.append(f"Sides: {result.sides}")
    lines.append("Distribution")
    lines.append("face | count | percentage")

    for face in sorted(result.distribution):
        count = result.distribution[face]
        pct = (count / result.total_rolls) * 100
        lines.append(f"{face} | {count} | {pct:.4f}")

    lines.append("")
    lines.append("Summary Statistics")
    lines.append(f"Mean: {result.mean:.6f}")
    lines.append(f"Variance: {result.variance:.6f}")
    lines.append(f"Std Dev: {result.std_dev:.6f}")
    return "\n".join(lines)


def format_json(result: SimulationResult) -> str:
    payload = {
        "total_rolls": result.total_rolls,
        "sides": result.sides,
        "distribution": {
            str(face): result.distribution[face] for face in sorted(result.distribution)
        },
        "mean": result.mean,
        "variance": result.variance,
        "std_dev": result.std_dev,
    }
    return json.dumps(payload, indent=2)


def format_csv(result: SimulationResult) -> str:
    output = StringIO()
    writer = csv.writer(output)

    # Distribution block (required face/count/percentage header).
    writer.writerow(["face", "count", "percentage"])
    for face in sorted(result.distribution):
        count = result.distribution[face]
        pct = (count / result.total_rolls) * 100
        writer.writerow([face, count, f"{pct:.4f}"])

    # Summary block so CSV still exposes the required aggregate metrics.
    writer.writerow([])
    writer.writerow(["summary_metric", "value"])
    writer.writerow(["total_rolls", result.total_rolls])
    writer.writerow(["sides", result.sides])
    writer.writerow(["mean", f"{result.mean:.6f}"])
    writer.writerow(["variance", f"{result.variance:.6f}"])
    writer.writerow(["std_dev", f"{result.std_dev:.6f}"])

    return output.getvalue().rstrip("\n")


def render_output(result: SimulationResult, output_format: str) -> str:
    if output_format == "text":
        return format_text(result)
    if output_format == "json":
        return format_json(result)
    if output_format == "csv":
        return format_csv(result)
    raise ValueError(f"Unsupported format: {output_format}")


def main(argv: Sequence[str] | None = None) -> int:
    args = argv if argv is not None else sys.argv[1:]

    try:
        parsed = parse_args(args)

        if parsed.parallel:
            # Spec allows a clear error when parallel mode is unsupported.
            raise ValueError("--parallel is not supported yet in Python implementation.")

        result = run_simulation(parsed.rolls, parsed.sides, parsed.seed)
        print(render_output(result, parsed.format))
        return 0

    except ValueError as exc:
        print(f"Input error: {exc}", file=sys.stderr)
        return 1
    except Exception as exc:  # Broad catch to map unexpected issues to code 2.
        print(f"Internal error: {exc}", file=sys.stderr)
        return 2


if __name__ == "__main__":
    raise SystemExit(main())

File: implementations/java/src/DiceLab.java
import java.util.Locale;
import java.util.Random;

public final class DiceLab {
    private enum OutputFormat {
        TEXT,
        JSON,
        CSV
    }

    private static final class CliOptions {
        int rolls = -1;
        int sides = 6;
        boolean hasSeed = false;
        long seed = 0L;
        OutputFormat format = OutputFormat.TEXT;
        boolean parallel = false;
        boolean showHelp = false;
    }

    private static final class SimulationResult {
        final int totalRolls;
        final int sides;
        final long[] distribution;
        final double mean;
        final double variance;
        final double stdDev;

        SimulationResult(int totalRolls, int sides, long[] distribution, double mean, double variance, double stdDev) {
            this.totalRolls = totalRolls;
            this.sides = sides;
            this.distribution = distribution;
            this.mean = mean;
            this.variance = variance;
            this.stdDev = stdDev;
        }
    }

    public static void main(String[] args) {
        try {
            CliOptions options = parseArgs(args);

            if (options.showHelp) {
                printHelp();
                System.exit(0);
                return;
            }

            if (options.parallel) {
                throw new IllegalArgumentException("--parallel is not supported yet in Java baseline.");
            }

            SimulationResult result = runSimulation(options);
            System.out.println(renderOutput(result, options.format));
            System.exit(0);
        } catch (IllegalArgumentException ex) {
            System.err.println("Input error: " + ex.getMessage());
            System.exit(1);
        } catch (Exception ex) {
            System.err.println("Internal error: " + ex.getMessage());
            System.exit(2);
        }
    }

    private static CliOptions parseArgs(String[] args) {
        CliOptions options = new CliOptions();

        // Single forward pass keeps parsing explicit and easy to debug.
        for (int index = 0; index < args.length; index++) {
            String argument = args[index];

            if ("--help".equals(argument)) {
                options.showHelp = true;
                continue;
            }

            if ("--parallel".equals(argument)) {
                options.parallel = true;
                continue;
            }

            if (index + 1 >= args.length) {
                throw new IllegalArgumentException("Missing value for " + argument + ".");
            }

            String value = args[++index];
            switch (argument) {
                case "--rolls" -> options.rolls = parseInt(value, "--rolls");
                case "--sides" -> options.sides = parseInt(value, "--sides");
                case "--seed" -> {
                    options.seed = parseLong(value, "--seed");
                    options.hasSeed = true;
                }
                case "--format" -> options.format = parseFormat(value);
                default -> throw new IllegalArgumentException("Unsupported argument: " + argument + ".");
            }
        }

        if (options.showHelp) {
            return options;
        }

        if (options.rolls <= 0) {
            throw new IllegalArgumentException("--rolls must be greater than 0.");
        }
        if (options.sides < 2) {
            throw new IllegalArgumentException("--sides must be greater than or equal to 2.");
        }

        return options;
    }

    private static int parseInt(String value, String flagName) {
        try {
            return Integer.parseInt(value);
        } catch (NumberFormatException ex) {
            throw new IllegalArgumentException(flagName + " must be an integer.");
        }
    }

    private static long parseLong(String value, String flagName) {
        try {
            return Long.parseLong(value);
        } catch (NumberFormatException ex) {
            throw new IllegalArgumentException(flagName + " must be an integer.");
        }
    }

    private static OutputFormat parseFormat(String value) {
        return switch (value) {
            case "text" -> OutputFormat.TEXT;
            case "json" -> OutputFormat.JSON;
            case "csv" -> OutputFormat.CSV;
            default -> throw new IllegalArgumentException("--format must be one of: text, json, csv.");
        };
    }

    private static SimulationResult runSimulation(CliOptions options) {
        // Determinism rule:
        // - with --seed: deterministic within this implementation
        // - without --seed: JVM-chosen seed from system entropy/time
        Random rng = options.hasSeed ? new Random(options.seed) : new Random();

        long[] counts = new long[options.sides];
        double total = 0.0;
        double totalSquares = 0.0;

        for (int i = 0; i < options.rolls; i++) {
            // nextInt(n) returns 0..n-1, so shift by +1.
            int value = rng.nextInt(options.sides) + 1;
            counts[value - 1]++;
            double asDouble = value;
            total += asDouble;
            totalSquares += asDouble * asDouble;
        }

        double mean = total / options.rolls;
        double variance = (totalSquares / options.rolls) - (mean * mean);
        if (variance < 0.0) {
            variance = 0.0;
        }

        return new SimulationResult(
            options.rolls,
            options.sides,
            counts,
            mean,
            variance,
            Math.sqrt(variance)
        );
    }

    private static String renderOutput(SimulationResult result, OutputFormat format) {
        return switch (format) {
            case TEXT -> formatText(result);
            case JSON -> formatJson(result);
            case CSV -> formatCsv(result);
        };
    }

    private static String formatText(SimulationResult result) {
        StringBuilder output = new StringBuilder();
        output.append("DiceLab Results\n");
        output.append("Total rolls: ").append(result.totalRolls).append("\n");
        output.append("Sides: ").append(result.sides).append("\n");
        output.append("Distribution\n");
        output.append("face | count | percentage\n");

        for (int face = 1; face <= result.sides; face++) {
            long count = result.distribution[face - 1];
            double percentage = (count / (double) result.totalRolls) * 100.0;
            output.append(String.format(Locale.US, "%d | %d | %.4f%n", face, count, percentage));
        }

        output.append("\nSummary Statistics\n");
        output.append(String.format(Locale.US, "Mean: %.6f%n", result.mean));
        output.append(String.format(Locale.US, "Variance: %.6f%n", result.variance));
        output.append(String.format(Locale.US, "Std Dev: %.6f", result.stdDev));
        return output.toString();
    }

    private static String formatJson(SimulationResult result) {
        // Manual JSON keeps face keys in predictable ascending order.
        StringBuilder output = new StringBuilder();
        output.append("{\n");
        output.append("  \"total_rolls\": ").append(result.totalRolls).append(",\n");
        output.append("  \"sides\": ").append(result.sides).append(",\n");
        output.append("  \"distribution\": {\n");

        for (int face = 1; face <= result.sides; face++) {
            long count = result.distribution[face - 1];
            output.append("    \"").append(face).append("\": ").append(count);
            if (face < result.sides) {
                output.append(",");
            }
            output.append("\n");
        }

        output.append("  },\n");
        output.append(String.format(Locale.US, "  \"mean\": %.6f,%n", result.mean));
        output.append(String.format(Locale.US, "  \"variance\": %.6f,%n", result.variance));
        output.append(String.format(Locale.US, "  \"std_dev\": %.6f%n", result.stdDev));
        output.append("}");
        return output.toString();
    }

    private static String formatCsv(SimulationResult result) {
        StringBuilder output = new StringBuilder();
        output.append("face,count,percentage\n");

        for (int face = 1; face <= result.sides; face++) {
            long count = result.distribution[face - 1];
            double percentage = (count / (double) result.totalRolls) * 100.0;
            output.append(String.format(Locale.US, "%d,%d,%.4f%n", face, count, percentage));
        }

        output.append("\nsummary_metric,value\n");
        output.append("total_rolls,").append(result.totalRolls).append("\n");
        output.append("sides,").append(result.sides).append("\n");
        output.append(String.format(Locale.US, "mean,%.6f%n", result.mean));
        output.append(String.format(Locale.US, "variance,%.6f%n", result.variance));
        output.append(String.format(Locale.US, "std_dev,%.6f", result.stdDev));
        return output.toString();
    }

    private static void printHelp() {
        System.out.println("Usage: dice-lab --rolls <int> [options]");
        System.out.println();
        System.out.println("Required:");
        System.out.println("  --rolls <int>      Total number of rolls (must be > 0)");
        System.out.println();
        System.out.println("Optional:");
        System.out.println("  --sides <int>      Number of die sides (must be >= 2, default: 6)");
        System.out.println("  --seed <int>       Seed for deterministic RNG");
        System.out.println("  --format <string>  Output format: text|json|csv (default: text)");
        System.out.println("  --parallel         Request parallel mode (unsupported in baseline)");
        System.out.println("  --help             Show this help message");
        System.out.println();
        System.out.println("Example:");
        System.out.println("  dice-lab --rolls 10000 --sides 20 --seed 42 --format json");
    }
}
